chain/network/src/ibf_peer_set.rs

```diff
@@ -0,0 +1,238 @@
+use near_primitives::network::PeerId;
+use std::collections::HashMap;
+use std::sync::{Arc, Mutex};
+
+use crate::ibf_set::IbfSet;
+use crate::routing::Edge;
+
+pub type SlotMapId = u64;
+
+#[derive(Default)]
+struct SlotMap {
+    id: u64,
+    id2e: HashMap<SlotMapId, SimpleEdge>,
+    e2id: HashMap<SimpleEdge, SlotMapId>,
+}
+
+impl SlotMap {
+    fn insert(&mut self, edge: &SimpleEdge) -> Option<SlotMapId> {
+        if self.e2id.contains_key(edge) {
+            return None;
+        }
+
+        let new_id = self.id as SlotMapId;
+        self.id += 1;
+
+        self.e2id.insert(edge.clone(), new_id);
+        self.id2e.insert(new_id, edge.clone());
+
+        Some(new_id)
+    }
+
+    fn get(&self, edge: &SimpleEdge) -> Option<SlotMapId> {
+        self.e2id.get(edge).cloned()
+    }
+
+    fn get_by_id(&self, id: &SlotMapId) -> Option<SimpleEdge> {
+        self.id2e.get(id).cloned()
+    }
+
+    fn pop(&mut self, edge: &SimpleEdge) -> Option<SlotMapId> {
+        if let Some(&id) = self.e2id.get(edge) {
+            self.e2id.remove(edge);
+            self.id2e.remove(&id);
+
+            return Some(id);
+        }
+        None
+    }
+}
+
+#[derive(Hash, Clone, Eq, PartialEq, Debug)]
+pub struct SimpleEdge {
+    pub key: (PeerId, PeerId),
+    pub nonce: u64,
+}
+
+#[derive(Default)]
+pub struct IbfPeerSet {
+    peers: HashMap<PeerId, Arc<Mutex<IbfSet<SimpleEdge>>>>,
```

Comment:
suggestion: replace `Arc<Mutex<IbfSet>>` with just `IbfSet`. If I understand correctly, we don't really need a mutex here, we just try to work-around a (admittedly non-trivial, [discussion](https://users.rust-lang.org/t/borrowing-error-when-downgrading-mut-t-to-t/64811)) borrow checker issue. I think the following should fix the issue:

```rust
                        // ibf_msg.seed
                        let ibf_set = if let Some(x) = self.routing_table.get_peer(&peer_id) {
                            x
                        } else {
                            self.routing_table.add_peer(peer_id.clone(), ibf_msg.seed);
                            // Do a second lookup with `&self`. Not elegant, but makes borrow checker happy 
                            // and a much smaller hack than a mutex.
                            self.routing_table.get_peer(&peer_id).unwrap()
                        };
```

Reply:
In the following code:
```

                        let ibf_set = if let Some(x) = self.routing_table.get_peer(&peer_id) {
                            x
                        } else {
                            self.routing_table.add_ibf_for_peer(peer_id.clone(), ibf_msg.seed)
                        };

///... and later
       self.routing_table.exchange_routing_tables_using_ibf(
                                    &peer_id,
                                    ibf_set.clone(),
                                    ibf_msg.ibf_level as usize,
                                    &ibf_msg.ibf,
                                    ibf_msg.seed,
                                );
```
In the second part of the code I'm using both borrowed `ibf_set`, and I'm using `self.routing_table.exchange_routing_tables_using_ibf`. 
I'm not sure how to easily avoid borrower checker, because we use `ibf_set` borrowed from `self.routing_table`, and `self` is needed for `       self.routing_table.exchange_routing_tables_using_ibf`

Reply:
I replaced `Arc` with `Rc` to indicate we use this data structure within a single thread. `Mutex` is not really needed, just to make `IbfSet` mutable.

Reply:
Then why not use `RefCell` instead?

Reply:
Ok, I changed it to `RefCell`.

Reply:
You appear to have changed it back to `Arc<Mutex<T>>`. What is the reason for that?

Reply:
I had to change it back, because I started using `IbfSet` across multiple thread. However, I did another refactoring, that's no longer the case. Now `IbfSet` is only used from `IbfRoutingTableExchangeActor`, so I changed it back again to `Rc<RefCell<T>>`